### PR TITLE
♿ Ensure amp-story-bookend images have an alt attribute

### LINKED ISSUE #32752: 🐛 `amp-story-bookend` preview/thumbnails missing `alt` attribute
### Current behavior
While this was not part of the evaluation we were asked to carry out, we noticed that `<amp-story-bookend>` allows the addition of images (e.g. to reference related stories/resources). It appears that currently there's no way to provide alternative text for these images (going by the documentation at least)

- https://amp.dev/documentation/guides-and-tutorials/start/visual_story/create_bookend/?format=stories
- https://amp.dev/documentation/components/amp-story-bookend/?format=stories

Looking at the example story here for instance https://amp.dev/documentation/examples/style-layout/amp_story_layouts/preview/?format=stories it seems the final output has an ``, but it lacks any `alt`. 

This leads to assistive technologies identifying that there is an unlabelled/unmarked image.

### Expected behavior/code
We'd recommend allowing authors the ability to add explicit `alt` for these bookend images. Failing that, the output `` should be given an empty `alt=""` to indicate that it is decorative only.

### Reproduction procedure
Please provide the steps to reproduce the problem:

1. Open the example story https://amp.dev/documentation/examples/style-layout/amp_story_layouts/preview/?format=stories
2. Navigate to the bookend
3. Verify that the images (which are currently broken?) use `` and lack an actual `alt` attribute

- Chrome / Windows

### PR DESCRIPTION
**Pull request by nickFalcone**
Images used on amp-story-bookend should be able to have a user-provided `alt` attribute. If the user does not provide an `alt` attribute, assign an [empty alt attribute](https://web.dev/image-alt/). 

- Create an `alt` key for Article, Portrait, and Landscape components
- Add tests to check that an `alt` attribute is added to the image (with text value if provided in JSON)
- Tested locally with Lighthouse Accessibility audit, Axe, and JAWS screen reader.

Resolves #32752

/cc @ampproject/wg-stories

Lighthouse, before:
before-lighthouse

Lighthouse, after:
after-lighthouse

Axe, before:
before-axe

Axe, after:
after-axe

### CODE REVIEW

File: examples/amp-story/bookendv1.json [hunk 1]
```diff
@@ -20,13 +20,15 @@
       "type": "small",
       "title": "This is an example article",
       "url": "/article.html",
-      "image": "http://placehold.it/500x600"
+      "image": "http://placehold.it/500x600",
+      "alt": "All images should receive an alt attribute, even if the value is an empty string or the alt key is not provided."
```
Comment (processprocess) [inline examples/amp-story/bookendv1.json:24]: We can reword this to mention default behavior:

```suggestion
      "alt": "Example alt text for the image. If alt is not specified it will default to an empty string."
```

File: examples/amp-story/bookendv1.json [hunk 2]
```diff
@@ -20,13 +20,15 @@
...
     },
     {
       "type": "small",
       "title": "This is an example article2",
       "url": "http://example.com/article.html",
-      "image": "http://placehold.it/256x128"
+      "image": "http://placehold.it/256x128",
+      "alt": ""
```
Comment (processprocess) [inline examples/amp-story/bookendv1.json:31]: Since this will happen by default we can remove it here.

File: extensions/amp-story/1.0/bookend/components/article.js [hunk 3]
```diff
@@ -77,6 +78,7 @@ export class ArticleComponent {
       domainName,
       type: articleJson['type'],
       title: articleJson['title'],
+      alt: articleJson['alt'],
```
Comment (processprocess) [inline extensions/amp-story/1.0/bookend/components/article.js:81]: This could be called `image-alt-text` since it's only applied to images.

File: extensions/amp-story/1.0/test/test-amp-story-bookend.js [hunk 4]
```diff
@@ -957,6 +967,47 @@ describes.fakeWin('amp-story-bookend', {win: {location}, amp: true}, (env) => {
...
+      const noAlt = small.buildElement(noAltComponent, win, {position: 0});
+      expect(noAlt.querySelector('img').alt).to.equal('');
+
+      const emptyAlt = small.buildElement(emptyAltComponent, win, {
+        position: 0,
+      });
+      expect(emptyAlt.querySelector('img').alt).to.equal('');
+
+      const textAlt = small.buildElement(textAltComponent, win, {position: 0});
+      expect(textAlt.querySelector('img').alt).to.equal('test');
```
Comment (processprocess) [inline extensions/amp-story/1.0/test/test-amp-story-bookend.js:1009]: Tests need to be focused on one concept. One `expect` per test.
Please separate these.

They could be something like:
`'should have an empty alt attribute on image if no alt is specified'`
`'should have an empty alt attribute on image if empty alt is specified'`
`'should propagate alt attribute if specified'`

I'm not sure that we need to test if the alt is empty if it was specified as empty. 
Do you think the first test covers that?

File: extensions/amp-story/1.0/test/test-amp-story-bookend.js [hunk 5]
```diff
@@ -1160,6 +1211,49 @@ describes.fakeWin('amp-story-bookend', {win: {location}, amp: true}, (env) => {
...
+
+      const emptyAlt = landscape.buildElement(emptyAltComponent, win, {
+        position: 0,
+      });
+      expect(emptyAlt.querySelector('img').alt).to.equal('');
+
+      const textAlt = landscape.buildElement(textAltComponent, win, {
+        position: 0,
+      });
+      expect(textAlt.querySelector('img').alt).to.equal('test');
```
Comment (processprocess) [inline extensions/amp-story/1.0/test/test-amp-story-bookend.js:1255]: Same comment as above.

File: extensions/amp-story/1.0/test/test-amp-story-bookend.js [hunk 6]
```diff
@@ -1353,5 +1447,48 @@ describes.fakeWin('amp-story-bookend', {win: {location}, amp: true}, (env) => {
...
+      const emptyAlt = portrait.buildElement(emptyAltComponent, win, {
+        position: 0,
+      });
+      expect(emptyAlt.querySelector('img').alt).to.equal('');
+
+      const textAlt = portrait.buildElement(textAltComponent, win, {
+        position: 0,
+      });
+      expect(textAlt.querySelector('img').alt).to.equal('test');
+    });
```
Comment (processprocess) [inline extensions/amp-story/1.0/test/test-amp-story-bookend.js:1492]: Same comment as above.

**Review by processprocess — COMMENTED** [review]
I have some comments but this is looking great @nickFalcone :) Thank you!

File: extensions/amp-story/1.0/test/test-amp-story-bookend.js (hunk 4 above)
Comment (nickFalcone) [reply extensions/amp-story/1.0/test/test-amp-story-bookend.js:1009]: I think you are right, testing alt is empty if it was specified empty seems redundant.  I'll separate these tests!

File: extensions/amp-story/1.0/bookend/components/article.js (hunk 3 above)
Comment (nickFalcone) [reply extensions/amp-story/1.0/bookend/components/article.js:81]: I tried renaming and ran into an issue with the following test:

https://github.com/ampproject/amphtml/blob/1b7080708cb0260689e24c8531d29354030ba700/extensions/amp-story/1.0/test/test-amp-story-bookend.js#L464-L500

I isolated the value of alt between `config` and `userJson` and see that the following passes: 

```js
expect(config.components[1].alt).to.equal(userJson.components[1]['image-alt-text']);

// expect(config.components[1]).to.deep.equal(userJson.components[1]);
```

I think `alt` vs `image-alt-text` might be causing the issue.  Do you have any thoughts on how to preserve the original test integrity with the rename?

Comment (processprocess) [reply extensions/amp-story/1.0/bookend/components/article.js:81]: Thanks for looking into this @nickFalcone
I looked into this a little deeper and `alt` is the naming convention used in our codebase. 
It's safe to stick with `alt` :)

Comment (nickFalcone) [reply extensions/amp-story/1.0/bookend/components/article.js:81]: Thanks @processprocess, sticking with `alt` all tests pass (including new separated tests) 😄

**Review by processprocess — APPROVED** [review]
Looks good! Thanks Nick :)
